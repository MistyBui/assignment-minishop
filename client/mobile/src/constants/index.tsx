export const ADD_PRODUCT = '+';
export type ADD_PRODUCT = typeof ADD_PRODUCT;

export const REDUCE_PRODUCT = '-';
export type REDUCE_PRODUCT = typeof REDUCE_PRODUCT;

export const DELETE_PRODUCT = 'x';
export type DELETE_PRODUCT = typeof DELETE_PRODUCT;

export const ORDER_CONFIRMATION = 'Order Confirmation';
export type ORDER_CONFIRMATION = typeof ORDER_CONFIRMATION;

export const ADD_TO_CART = 'Add To Cart';
export type ADD_TO_CART = typeof ADD_TO_CART;

export const MY_ORDERS = 'My Orders';
export type MY_ORDERS = typeof MY_ORDERS;

export const PRODUCTS = 'Products';
export type PRODUCTS = typeof PRODUCTS;

export const ORDERS = 'Orders';
export type ORDERS = typeof ORDERS;

export const ORDER_DETAILS = 'OrderDetails';
export type ORDER_DETAILS = typeof ORDER_DETAILS;
