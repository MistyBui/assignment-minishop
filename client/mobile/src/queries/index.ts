import { gql } from '@apollo/client';

export const getProductsQuery = gql(`
  query getProducts {
    products {
      ean
      imageUrl
      name
      price
    }
  }
`);

export const ordersQuery = gql(`
query getOrders($customerId: ID!) {
  orders(customerId: $customerId) {
    orderId
    customerId
    timestamp
    products {
      amount
      ean
    }
    totalSum
  }
}`);
