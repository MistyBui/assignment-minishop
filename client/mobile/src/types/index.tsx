export type Product = {
  name: string;
  ean: string;
  imageUrl: string;
  price: number;
};

export type CartItem = {
  product: Product;
  quantity: number;
};

export type CartContextType = {
  cart: CartItem[];
  addToCart: (product: Product) => void;
  reduceFromCart: (product: Product) => void;
  removeFromCart: (ean: string) => void;
  clearCart: () => void;
};

export interface OrderCard {
  orderId: string;
  timestamp: string;
  totalSum: number;
}
