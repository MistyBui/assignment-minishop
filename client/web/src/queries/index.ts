import { gql } from 'graphql-tag';

export const getProductsQuery = gql(`
  query getProducts {
    products {
      ean
      imageUrl
      name
      price
    }
  }
`);

export const ordersQuery = gql(`
  query getOrders($customerId: ID!) {
    orders(customerId: $customerId) {
      orderId
      customerId
      timestamp
      products {
        amount
        ean
      }
      totalSum
    }
  }`);
