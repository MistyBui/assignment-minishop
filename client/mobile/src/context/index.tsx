import React, { createContext, useContext, useState } from 'react';
import { CartContextType, CartItem, Product } from 'types';

interface Props {
  children: React.ReactNode;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

export const useCart = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error('useCart must be used within a CartProvider');
  }
  return context;
};

export const CartProvider: React.FC<Props> = ({ children }) => {
  const [cart, setCart] = useState<CartItem[]>([]);

  const addToCart = (product: Product) => {
    const existingItem = cart.find(item => item.product.ean === product.ean);
    if (existingItem) {
      setCart(prevCart =>
        prevCart.map(item =>
          item.product.ean === product.ean
            ? { ...item, quantity: item.quantity + 1 }
            : item,
        ),
      );
    } else {
      setCart(prevCart => [...prevCart, { product, quantity: 1 }]);
    }
  };

  const reduceFromCart = (product: Product) => {
    setCart(prevCart =>
      prevCart
        .map(item =>
          item.product.ean === product.ean
            ? { ...item, quantity: item.quantity - 1 }
            : item,
        )
        .filter(item => item.quantity > 0),
    );
  };

  const removeFromCart = (ean: string) => {
    setCart(prevCart => prevCart.filter(item => item.product.ean !== ean));
  };

  const clearCart = () => {
    setCart([]);
  };

  const contextValue: CartContextType = {
    cart,
    addToCart,
    reduceFromCart,
    removeFromCart,
    clearCart,
  };

  return (
    <CartContext.Provider value={contextValue}>{children}</CartContext.Provider>
  );
};
