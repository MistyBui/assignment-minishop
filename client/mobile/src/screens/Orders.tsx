import type { GetOrdersQuery } from '../../../../server/src/generated/graphql';

import React, { FC } from 'react';
import { ScrollView, StyleSheet, Text, View } from 'react-native';
import { useQuery } from '@apollo/client';

import Header from '../components/Header';
import OrderSummary from '../components/OrderSummary';
import { OrderCard } from '../types';
import { ordersQuery } from '../queries';
import { ORDERS } from '../constants';

const Orders: FC = (): JSX.Element => {
  const { loading, data } = useQuery<GetOrdersQuery>(ordersQuery, {
    variables: { customerId: 'customer-1' },
  });

  return (
    <View style={styles.wrapper}>
      <Header>{ORDERS}</Header>
      <ScrollView>
        {loading ? (
          <Text>Loading orders...</Text>
        ) : (
          data?.orders.map((order: OrderCard) => (
            <OrderSummary
              key={order.orderId}
              orderId={order.orderId}
              timestamp={order.timestamp}
              totalSum={order.totalSum}
            />
          ))
        )}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    padding: 16,
    paddingTop: 48,
  },
});

export default Orders;
