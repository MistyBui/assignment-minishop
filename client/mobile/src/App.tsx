import React, { createRef } from 'react';
import { createNativeStackNavigator } from '@react-navigation/native-stack';

import Products from './screens/Products';
import Orders from './screens/Orders';
import OrderDetails from './screens/OrderDetails';
import { CartProvider } from './context/index';
import { PRODUCTS, ORDERS, ORDER_DETAILS } from './constants';
import {
  NavigationContainer,
  NavigationContainerRef,
} from '@react-navigation/native';

export type RootStackParamList = {
  Products: undefined;
  Orders: undefined;
  OrderDetails: undefined;
};

const Stack = createNativeStackNavigator<RootStackParamList>();

export const navigationRef = createRef<NavigationContainerRef<any>>();

export const navigate = (name: string, params?: object) => {
  navigationRef.current?.navigate(name, params);
};

const App = () => {
  return (
    <CartProvider>
      <NavigationContainer independent={true} ref={navigationRef}>
        <Stack.Navigator
          initialRouteName={PRODUCTS}
          screenOptions={{ headerShown: false }}>
          <Stack.Screen name={PRODUCTS} component={Products} />
          <Stack.Screen name={ORDERS} component={Orders} />
          <Stack.Screen name={ORDER_DETAILS} component={OrderDetails} />
        </Stack.Navigator>
      </NavigationContainer>
    </CartProvider>
  );
};

export default App;
