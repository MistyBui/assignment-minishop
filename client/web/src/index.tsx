import React from 'react';
import { createRoot } from 'react-dom/client';
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client';
import { loadErrorMessages, loadDevMessages } from '@apollo/client/dev';

import OrderDetails from './screens/OrderDetails';
import Orders from './screens/Orders';
import Store from './screens/Products';
import { CartProvider } from './context';

const gqlClient = new ApolloClient({
  uri: 'http://localhost:4000',

  cache: new InMemoryCache(),
});
loadDevMessages();
loadErrorMessages();

const router = createBrowserRouter([
  {
    path: '/',
    element: <Store />,
  },
  {
    path: 'orders/',
    element: <Orders />,
  },
  {
    path: 'order-details/',
    element: <OrderDetails />,
  },
]);

const container = document.getElementById('app');
const root = createRoot(container);
root.render(
  <ApolloProvider client={gqlClient}>
    <CartProvider>
      <RouterProvider router={router} />
    </CartProvider>
  </ApolloProvider>,
);
