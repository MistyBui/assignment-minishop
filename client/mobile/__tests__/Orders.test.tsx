import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import Orders from '../src/screens/Orders';
import { ordersQuery } from '../src/queries';
import { NavigationContainer } from '@react-navigation/native';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
  useRoute: jest.fn(),
}));

const mocks = [
  {
    request: {
      query: ordersQuery,
      variables: { customerId: 'customer-1' },
    },
    result: {
      data: {
        orders: [
          {
            orderId: 'order-1',
            timestamp: '2023-09-22T12:00:00Z',
            totalSum: 100,
          },
        ],
      },
    },
  },
];

describe('Orders Screen', () => {
  it('renders loading state initially', () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <NavigationContainer>
          <Orders />
        </NavigationContainer>
      </MockedProvider>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders orders when data is loaded', async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <NavigationContainer>
          <Orders />
        </NavigationContainer>
        ;
      </MockedProvider>,
    );

    // Wait for the Apollo queries to resolve
    await new Promise(resolve => setTimeout(resolve, 0));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
