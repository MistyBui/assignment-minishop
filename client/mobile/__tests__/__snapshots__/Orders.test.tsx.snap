// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders Screen renders loading state initially 1`] = `
<View
  style={
    {
      "padding": 16,
      "paddingTop": 48,
    }
  }
>
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "marginStart": 10,
        }
      }
    >
      Orders
    </Text>
  </View>
  <RCTScrollView>
    <View>
      <Text>
        Loading orders...
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Orders Screen renders orders when data is loaded 1`] = `null`;
